### PR TITLE
feat: refine region state checks and handle stalled requests

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

refine region state checks and handle stalled requests

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

## Summary by CodeRabbit

- **New Features**
	- Introduced `NotifyRegionChangeResultListener` for enhanced region change notifications.
	- Added asynchronous test for handling concurrent alter requests.
	- New methods for improved error handling in region management, including `get_region_or` and `writable_region_or`.
	- Enhanced management of stalled requests with new methods for rejection and handling.

- **Bug Fixes**
	- Improved error handling and state management in region edit and change requests.

- **Documentation**
	- Updated method signatures to reflect new constraints and functionalities.

### CODE REVIEW

**Comment by WenyXu** [comment]
@coderabbitai summary

File: src/mito2/src/worker.rs [hunk 1]
```diff
@@ -579,7 +580,7 @@ type RequestBuffer = Vec<WorkerRequest>;
 #[derive(Default)]
 pub(crate) struct StalledRequests {
     /// Stalled requests.
-    pub(crate) requests: Vec<SenderWriteRequest>,
+    pub(crate) requests: HashMap<RegionId, Vec<SenderWriteRequest>>,
```
Comment (fengjiachun) [inline src/mito2/src/worker.rs:583]: `HashMap<RegionId, (estimated_size: usize, Vec<SenderWriteRequest>)>,`

We can avoid recalculating `estimated_size` when the remove operation occurs.

Comment (evenyag) [reply src/mito2/src/worker.rs:583]: Maybe we can store the estimated size for each region.

File: src/mito2/src/worker/handle_manifest.rs [hunk 2]
```diff
@@ -294,6 +302,9 @@ impl<S> RegionWorkerLoop<S> {
 
         // Sets the region as writable.
         region.switch_state_to_writable(RegionLeaderState::Altering);
+        // Handles the stalled requests.
+        self.handle_region_stalled_requests(&change_result.region_id)
+            .await;
```
Comment (evenyag) [inline src/mito2/src/worker/handle_manifest.rs:307]: Handle stalled requests after sending result to the sender so we won't block the waiter for too long.

File: src/mito2/src/worker/handle_manifest.rs [hunk 3]
```diff
@@ -74,7 +75,7 @@ impl RegionEditQueue {
     }
 }
 
-impl<S> RegionWorkerLoop<S> {
+impl<S: LogStore> RegionWorkerLoop<S> {
```
Comment (evenyag) [inline src/mito2/src/worker/handle_manifest.rs:78]: Maybe we can only add the trait bound to `handle_manifest_region_change_result()`

File: src/mito2/src/worker/handle_alter.rs [hunk 4]
```diff
@@ -32,7 +33,7 @@ use crate::region::MitoRegionRef;
 use crate::request::{DdlRequest, OptionOutputTx, SenderDdlRequest};
 use crate::worker::RegionWorkerLoop;
 
-impl<S> RegionWorkerLoop<S> {
+impl<S: LogStore> RegionWorkerLoop<S> {
```
Comment (evenyag) [inline src/mito2/src/worker/handle_alter.rs:36]:
```suggestion
impl<S> RegionWorkerLoop<S> {
```

File: src/mito2/src/worker.rs [hunk 5]
```diff
@@ -579,20 +580,37 @@ type RequestBuffer = Vec<WorkerRequest>;
...
+
+    /// Pushes a stalled request to the buffer.
+    pub(crate) fn push(&mut self, req: SenderWriteRequest) -> usize {
+        let (size, requests) = self.requests.entry(req.request.region_id).or_default();
+        let req_size = req.request.estimated_size();
+
+        *size += req_size;
+        requests.push(req);
+        req_size
+    }
```
Comment (evenyag) [inline src/mito2/src/worker.rs:604]: We can update `estimated_size` in `push()`

**Review by evenyag — APPROVED** [review]
Others LGTM

File: src/mito2/src/worker.rs [hunk 6]
```diff
@@ -579,20 +580,36 @@ type RequestBuffer = Vec<WorkerRequest>;
 #[derive(Default)]
 pub(crate) struct StalledRequests {
     /// Stalled requests.
-    pub(crate) requests: Vec<SenderWriteRequest>,
+    pub(crate) requests: HashMap<RegionId, (usize, Vec<SenderWriteRequest>)>,
```
Comment (killme2008) [inline src/mito2/src/worker.rs:583]: Better to document the key and value in the map.

**Review by fengjiachun — APPROVED** [review]
LGTM

**Pull request merged**